packages/lexical-playground/src/plugins/TableOfContentsPlugin.tsx

```diff
@@ -0,0 +1,159 @@
+    const elementYPosition = element?.getClientRects()[0].y;
+    return elementYPosition > 0.26 && elementYPosition < 9;
+  }
+  function isElementAboveViewport(element: HTMLElement): boolean {
+    const elementYPosition = element?.getClientRects()[0].y;
+    return elementYPosition <= 0;
+  }
+  function isElementBelowTheTopOfThePage(element: HTMLElement): boolean {
+    const elementYPosition = element?.getClientRects()[0].y;
+    return elementYPosition > 9;
+  }
+
+  function scrollCallback() {
+    if (tableOfContents.length !== 0) {
+      let currentHeading = editor.getElementByKey(
+        tableOfContents[selectedIndex.current][0],
+      );
+      if (currentHeading !== null) {
+        if (isElementAboveViewport(currentHeading)) {
+          //On natural scroll, user is scrolling down
+          while (
+            currentHeading !== null &&
+            isElementAboveViewport(currentHeading) &&
+            selectedIndex.current < tableOfContents.length - 1
+          ) {
+            const nextHeading = editor.getElementByKey(
+              tableOfContents[selectedIndex.current + 1][0],
+            );
+            if (
+              nextHeading !== null &&
+              (isElementAtTheTopOfThePage(nextHeading) ||
+                isElementAboveViewport(nextHeading))
+            ) {
+              const nextHeadingKey =
+                tableOfContents[++selectedIndex.current][0];
+              setSelectedKey(nextHeadingKey);
+              currentHeading = nextHeading;
+              break;
+            } else {
+              currentHeading = nextHeading;
+            }
+          }
+        } else if (isElementBelowTheTopOfThePage(currentHeading)) {
+          //On natural scroll, user is scrolling up
+          while (
+            currentHeading !== null &&
+            isElementBelowTheTopOfThePage(currentHeading) &&
+            selectedIndex.current > 0
+          ) {
+            const prevHeading = editor.getElementByKey(
+              tableOfContents[selectedIndex.current - 1][0],
+            );
+            if (
+              prevHeading !== null &&
+              (isElementAtTheTopOfThePage(prevHeading) ||
+                isElementAboveViewport(prevHeading))
+            ) {
+              const prevHeadingKey =
+                tableOfContents[--selectedIndex.current][0];
+              setSelectedKey(prevHeadingKey);
+              break;
+            } else {
+              currentHeading = prevHeading;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  useEffect(() => {
+    let timerId: ReturnType<typeof setTimeout>;
+
+    function debounceFunction(func: () => void, delay: number) {
+      clearTimeout(timerId);
+      timerId = setTimeout(func, delay);
+    }
+
+    function onScroll(): void {
+      debounceFunction(scrollCallback, 10);
+    }
+
+    document.addEventListener('scroll', onScroll);
+    return () => document.removeEventListener('scroll', onScroll);
+  });
```

Comment:
Also, we may want to look into the 3 array optimization but we can do this in a separate PR
```suggestion
  }, [...]);
```